File: web/src/lib/api-keys.ts
import { apiClient } from '@/lib/api'

export type ApiKeyStatus = 'active' | 'revoked'

export interface ApiKey {
  id: string           // UUID
  name: string
  key_prefix: string   // e.g. "flm_a1b2c3d4"
  status: ApiKeyStatus
  expires_at: string | null   // ISO timestamp
  last_used_at: string | null // ISO timestamp
  created_at: string | null   // ISO timestamp
}

export interface ApiKeyCreated extends ApiKey {
  full_key: string    // shown ONCE — never re-fetchable
}

export interface ApiKeyListResponse {
  keys: ApiKey[]
  total: number
}

export interface CreateApiKeyRequest {
  name: string
  expires_at?: string | null  // ISO timestamp
}

// All endpoints are under /internal/keys (JWT-protected on backend)
export async function listApiKeys(): Promise<ApiKeyListResponse> {
  const res = await apiClient<{ status: string; data: ApiKeyListResponse }>('/keys')
  return res.data
}

export async function createApiKey(req: CreateApiKeyRequest): Promise<ApiKeyCreated> {
  const res = await apiClient<{ status: string; data: ApiKeyCreated; message: string }>('/keys', {
    method: 'POST',
    body: JSON.stringify(req),
  })
  return res.data
}

export async function revokeApiKey(id: string): Promise<ApiKey> {
  const res = await apiClient<{ status: string; data: ApiKey; message: string }>(`/keys/${id}`, {
    method: 'DELETE',
  })
  return res.data
}

File: backend/src/internal/schema/api_keys.py
import uuid
from datetime import datetime

from pydantic import BaseModel

from src.model.api import ApiKeyStatus


class CreateApiKeyRequest(BaseModel):
    """Request to create a new API key — name, optional expiry."""
    name: str
    expires_at: datetime | None = None


class UpdateApiKeyRequest(BaseModel):
    """Request to update an API key — name and/or expiry."""
    name: str | None = None
    expires_at: datetime | None = None


class ApiKeyResponse(BaseModel):
    """API key response (full_key excluded — only returned once at creation)."""
    model_config = {"from_attributes": True}

    id: uuid.UUID
    name: str
    key_prefix: str
    status: ApiKeyStatus
    expires_at: datetime | None = None
    last_used_at: datetime | None = None
    created_at: datetime | None = None


# Extends ApiKeyResponse with full_key because the raw key is only shown once at creation
class ApiKeyCreatedResponse(ApiKeyResponse):
    full_key: str


class ApiKeyListResponse(BaseModel):
    keys: list[ApiKeyResponse]
    total: int
